src/borg/helpers/fs.py

```diff
@@ -164,12 +164,39 @@ def dir_is_tagged(path, exclude_caches, exclude_if_present):
     return tag_names
 
 
-_safe_re = re.compile(r"^((\.\.)?/+)+")
+def make_path_safe(path):
+    """
+    Make path safe by making it relative and normalized
 
+    `path` is normalized by making it relative, removing
+    consecutive slashes (e.g. '//'), remove '.' elements,
+    and removing trailing slashes.
 
-def make_path_safe(path):
-    """Make path safe by making it relative and local"""
-    return _safe_re.sub("", path) or "."
+    For reasons of security, a ValueError is raised should
+    `path` contain any '..' elements.
+    """
+    path = path.lstrip("/")
+    if path.startswith("../") or "/../" in path or path.endswith("/..") or path == "..":
+        raise ValueError(f"unexpected '..' element in path {path!r}")
+    path = os.path.normpath(path)
+    return path
+
+
+_dotdot_re = re.compile(r"^(\.\./)*")
+
+
+def remove_dotdot_prefixes(path):
+    """
+    Remove leading '../'s at the begging of `path`. Additionally,
+    the path is made relative.
+
+    `path` is expected to be normalized already (e.g. via `os.path.normpath()`).
+    """
+    path = path.lstrip("/")
+    path = _dotdot_re.sub("", path)
+    if path in ["", ".."]:
+        return "."
+    return path
```

Comment:
Which kind of paths should be accepted/rejected? I assume we still want to allow users to specify absolute paths and simple relative paths like a/b/c or ./a/b/c. So, are you suggesting to refuse any ../some/path?

Note that the regex starts with ^ and we only remove prefixes from the path.

Reply:
Guess we can approach this from 2 perspectives:

a) **what does your PR change compared to the existing code**

for archiving items the new `remove_dotdot_prefixes` is quite close to what `make_path_safe` did, except that it removes leading slashes (which we must do somewhere) and handles the special cases of "nothing left" ("" and ".."), which is also fine (except that a mere "." as archived path is not really useful for extraction)

b) **what do we really want it to be**

usually borg recurses starting from the recursion roots (which we can normalize first) and then only pretty normal paths will be generated by the recursion. for this, we do not need special path processing per item.

when fed with a paths list via stdin (and not using borg's recursor), borg does not have control over what's coming in from there, but borg is also are not required to accept too crappy stuff and still make great sense from it (the admin or tool feeding that list into borg can be expected to provide reasonable paths), borg instead could skip invalid paths with an error msg.

```
>>> normpath("recursion_root/foo/../bar")
'recursion_root/bar'  # definitely valid, it's inside the recursion root

>>> normpath("recursion_root/foo/../../bar")
'bar'  # slightly crappy, but we could archive this correctly (sibling of recursion root)

>>> normpath("recursion_root/foo/../../../bar")
'../bar'  # we can not archive this as is (../ unwanted due to security) and if we strip the ../ we change the meaning.
```

Reply:
> when fed with a paths list via stdin (and not using borg's recursor), borg does not have control over what's coming in from there, but borg is also not required to accept too crappy stuff and still make great sense from it (the admin or tool feeding that list into borg can be expected to provide reasonable paths), borg instead could skip invalid paths with an error msg.

I, personally, wouldn't give this too much consideration. Seems unlikely to me that people will end up providing such crappy paths to Borg. Paths with './' or '//', sure, but something like "root/foo/../../../bar" seems rather unlikely to me.

Reply:
Handling of '.' we should perhaps be looked at separately. There are some cases where it isn't handled ideally:

```
user@dev-borg:~$ borg list 85
-rw-r--r-- root   root        905 Sun, 2022-04-10 12:08:45  etc/fstab
user@dev-borg:~$ borg create --paths-from-stdin 86 <<<$'/etc/fstab\n'
user@dev-borg:~$ borg list 86
-rw-r--r-- root   root        905 Sun, 2022-04-10 12:08:45  etc/fstab
drwx------ user   user          0 Sun, 2022-11-13 18:19:19  .
```

Empty line at end is interpreted as '.' somehow.

Reply:
yeah, likely.

about archiving "." directory making sense: yes, i somehow agree, but how would you ever extract that? borg usually expects nothing being in the way, so it usually rmdirs target and then extracts target. not sure whether that would work with target == ".".

Reply:
Hmm, thought more about this, and guess I have to correct myself:
- "losing information" --> yes, we remove "../../../", so we lose that. **But:** that is not super useful as borg does not track the cwd. The admin might know and it might be reconstructable from scripting etc., but it is not much we lose.
- Also, and maybe more important: by using some "../" to point to parent dirs, the resulting paths usually **get longer**, so they gain information. e.g. if cwd is "/etc", if we refer to "passwd" we just have that. if we refer to "../etc/passwd" and then remove "../" from that, we have "etc/passwd", which is **more**.